src/main/java/io/netty/incubator/codec/http3/QpackEncoderDynamicTable.java

```diff
@@ -0,0 +1,390 @@
+     *          Insertion Index                 Draining Index  Dropping Index
+     */
+
+    /**
+     * Head of the entries, such that {@link HeaderEntry#index} is the {@code droppingIndex}.
+     */
+    private final HeaderEntry head;
+
+    /**
+     * Pointer before which entries are marked for eviction post {@link #incrementKnownReceivedCount(int)}.
+     * {@link HeaderEntry#index} is the {@code drainingIndex}.
+     */
+    private HeaderEntry drain;
+
+    /**
+     * Pointer before which entries are {@link #incrementKnownReceivedCount(int) acknowledged} but not yet marked for
+     * eviction.
+     */
+    private HeaderEntry ackTail;
+
+    /**
+     * Tail of the entries, such that {@link HeaderEntry#index} is the {@code insertionIndex}.
+     */
+    private HeaderEntry tail;
+
+    QpackEncoderDynamicTable() {
+        this(16, 10);
+    }
+
+    QpackEncoderDynamicTable(int arraySizeHint, int expectedFreeCapacityPercentage) {
+        // Enforce a bound of [2, 128] because hashMask is a byte. The max possible value of hashMask is one less
+        // than the length of this array, and we want the mask to be > 0.
+        fields = new HeaderEntry[findNextPositivePowerOfTwo(max(2, min(arraySizeHint, 128)))];
+        hashMask = (byte) (fields.length - 1);
+        head = new HeaderEntry(-1, EMPTY_STRING, EMPTY_STRING, 0, null);
+        this.expectedFreeCapacityPercentage = expectedFreeCapacityPercentage;
+        resetIndicesToHead();
+    }
+
+    int add(CharSequence name, CharSequence value, long headerSize) {
+        if (capacity - size < headerSize) {
+            return -1;
+        }
+
+        int h = AsciiString.hashCode(name);
+        int i = index(h);
+        HeaderEntry old = fields[i];
+        HeaderEntry e = new HeaderEntry(h, name, value, nextIndex(), old);
+        fields[i] = e;
+        e.addNextTo(tail);
+        tail = e;
+        size += headerSize;
+
+        ensureFreeCapacity();
+        return e.index;
+    }
+
+    void acknowledgeInsertCount(int requiredInsertCount) throws QpackException {
+        if (requiredInsertCount <= 0) {
+            throw INVALID_REQUIRED_INSERT_COUNT_INCREMENT;
+        }
+        // requiredInsertCount() adds 1 to the count, so adjusting the same.
+        requiredInsertCount--;
+        for (;;) {
+            if (ackTail.index == requiredInsertCount) {
+                ackTail.refCount--;
+                evictUnreferencedEntries();
+                return;
+            }
+            if (ackTail.next == null) {
+                break;
+            }
+            ackTail = ackTail.next;
+        }
+
+        // We have reached the end of the linked list so the index was invalid and hence the connection should be
+        // closed.
+        // https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#section-4.4
+        throw new IllegalArgumentException("Invalid value for required insert count: " + (requiredInsertCount + 1));
```

Comment:
I would just remove the `break` above in the `for` loop and throw directly there. I think thats easier to read and cleaner